project/src/main/java/com/google/sps/servlets/SignUpServlet.java

```diff
@@ -39,37 +46,74 @@
-      userEntity.setProperty("metric", "fahrenheit");
-      userEntity.setProperty("admin", false);
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String name = request.getParameter("fname") + " " + request.getParameter("lname");
+        String email = request.getParameter("email").toLowerCase();
+        String birthdate = request.getParameter("birthdate");
+        long studentId = Long.parseLong(request.getParameter("studentID"));
+        String sex = request.getParameter("sex");
+        String school = request.getParameter("school");
+        String phone = request.getParameter("phone");
+        String password = hashPassword(request.getParameter("password").toCharArray());
 
+        Query query = new Query("User");
 
-      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-      datastore.put(userEntity);
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+        ArrayList<String> users = new ArrayList<String>();
+        for(Entity entity:results.asIterable()){
+            if(entity.getProperty("email").equals(email)){
+                response.sendRedirect("/login.html");
+                return;
+            }
+        }
 
-      User user = new User(userEntity.getKey().getId(), name, email, birthdate, studentId, sex, school, phone, "fahrenheit", false);
-      Gson gson = new Gson();
-      response.setContentType("application/json;");
-      response.getWriter().println(gson.toJson(user));
+        Entity userEntity = new Entity("User");
+        userEntity.setProperty("name", name);
+        userEntity.setProperty("email", email);
+        userEntity.setProperty("birthdate", birthdate);
+        userEntity.setProperty("studentId", studentId);
+        userEntity.setProperty("sex", sex);
+        userEntity.setProperty("school", school);
+        userEntity.setProperty("phone", phone);
+        userEntity.setProperty("metric", "fahrenheit");
+        userEntity.setProperty("admin", false);
+        userEntity.setProperty("password", password);
 
-      response.sendRedirect("/dashboard.html");
-  }
+        datastore.put(userEntity);
+
+        user = new User(userEntity.getKey().getId(), name, email, password, birthdate, studentId, sex, school, phone, "fahrenheit", false);
+        Gson gson = new Gson();
+        response.setContentType("application/json;");
+        response.getWriter().println(gson.toJson(user));
+
+        response.sendRedirect("/dashboard.html");
+    }
+
+        @Override
+        public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException{
+            Gson gson = new Gson();
+            response.setContentType("application/json;");
+            response.getWriter().println(gson.toJson(user));
+        }
+
+
+    private String hashPassword(final char[] password){
```

Comment:
Some methods are similar to the methods LogInServlet.java, could you move common methods in separate lib and reuse it in both files please

Reply:
done.